### PR TITLE
check to ensure $value is an array in WP_List_Util::pluck()

### PR DESCRIPTION
**Pull request by afragen**
Add a check to the `else` as `elseif` to ensure `$value` is an array when it is treated as such.

Trac ticket: https://core.trac.wordpress.org/ticket/56650

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-list-util.php [hunk 1]
```diff
@@ -166,9 +166,17 @@ public function pluck( $field, $index_key = null ) {
 			foreach ( $this->output as $key => $value ) {
 				if ( is_object( $value ) ) {
 					$newlist[ $key ] = $value->$field;
-				} else {
+				} elseif ( is_array( $value ) ) {
 					$newlist[ $key ] = $value[ $field ];
 				}
+
+				if ( ! ( is_object( $value ) || is_array( $value ) ) ) {
```
Comment (costdev) [inline src/wp-includes/class-wp-list-util.php]:
```suggestion
				if ( ! is_object( $value ) && ! is_array( $value ) ) {
```
IMO, this is a little easier to read. Just a suggestion though. Same suggestion for the later instance.

File: src/wp-includes/class-wp-list-util.php [hunk 2]
```diff
@@ -166,9 +166,17 @@ public function pluck( $field, $index_key = null ) {
...
+					_doing_it_wrong(
+						__FUNCTION__,
+						__( 'Value for `$this->output` must be either an object or an array for this function.' ),
```
Comment (costdev) [inline src/wp-includes/class-wp-list-util.php]:
```suggestion
						__( "Values for the input array must be either objects or arrays.' ),
```
Same suggestion in the later instance.

**Review by costdev — CHANGES_REQUESTED** [review]
Just some minor suggestions. Still needs unit tests - I'm happy to take a look at those.

File: src/wp-includes/class-wp-list-util.php (hunk 1 above)
Comment (costdev) [inline src/wp-includes/class-wp-list-util.php]: Hmm...

Starting to wonder if an `else` is more appropriate for the `_doing_it_wrong()` condition. That saves on some checks.

i.e.
Currently:
- Good object data: `if is object` + `if not object and not array` (2 - one more than `trunk`)
- Good array data: `if is object` + `else if is array` + `if not object and not array` (3 - one more than `trunk`)
- Bad data: `if is object` + `else if is array` + `if not object and not array` (3)

With `else`:
- Good object data: `if is object` (1 - no change from `trunk`)
- Good array data: `if is object` + `else if is array` (2 - no change from `trunk`)
- Bad data: `if is object` + `else if is array` + `else` (3ish)

**Comment by afragen** [comment]
@costdev I think I've resolved the above issues.

**Review by costdev — APPROVED** [review]
LGTM :+1:

File: tests/phpunit/tests/functions/wpListPluck.php [hunk 3]
```diff
@@ -350,4 +350,63 @@ public function data_test_wp_list_pluck() {
...
+	 * @covers WP_List_Util::pluck
+	 *
+	 * @dataProvider data_wp_list_pluck_should_throw_doing_it_wrong_with_invalid_input
+	 *
+	 * @expectedIncorrectUsage WP_List_Util::pluck
+	 *
+	 * @param array $input An invalid input array.
+	 */
+	public function test_wp_list_pluck_should_throw_doing_it_wrong_with_invalid_input( $input ) {
+		$this->assertSame( array(), wp_list_pluck( $input, 'a_field' ) ); // I think...
```
Comment (audrasjb) [inline tests/phpunit/tests/functions/wpListPluck.php]: Is the "// I think..." comment intended?

File: tests/phpunit/tests/functions/wpListPluck.php [hunk 4]
```diff
@@ -350,4 +350,63 @@ public function data_test_wp_list_pluck() {
...
+	 * @covers WP_List_Util::pluck
+	 *
+	 * @dataProvider data_wp_list_pluck_should_throw_doing_it_wrong_with_invalid_input
+	 *
+	 * @expectedIncorrectUsage WP_List_Util::pluck
+	 *
+	 * @param array $input An invalid input array.
+	 */
+	public function test_wp_list_pluck_should_throw_doing_it_wrong_with_index_key_and_invalid_input( $input ) {
+		$this->assertSame( array(), wp_list_pluck( $input, 'a_field', 'an_index_key' ) ); // I think...
```
Comment (audrasjb) [inline tests/phpunit/tests/functions/wpListPluck.php]: Is the "// I think..." comment intended?

File: tests/phpunit/tests/functions/wpListPluck.php (hunk 3 above)
Comment (afragen) [reply tests/phpunit/tests/functions/wpListPluck.php]: @costdev I think these are for you 😉

Comment (costdev) [reply tests/phpunit/tests/functions/wpListPluck.php]: ...woops! Nope, that was left over from pseudo during a DM chat about the tests. :sweat_smile:

**Comment by SergeyBiryukov** [comment]
Thanks for the PR! Merged in [r55423](https://core.trac.wordpress.org/changeset/55423).

Note: As `wp_list_pluck()` is just a wrapper and the actual source code change was in `WP_List_Util::pluck()`, I have moved the tests to the class where all the other tests for `WP_List_Util` are located.

**Pull request closed without merging**